keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,153 @@
+This KEP proposes to implement readonly & mutating `capacity` for the apiserver; readonly & mutating `limit` and `request` of requests for different users and groups, considering different physical resource consumption by particular operations (e.g. get single pod vs. get all pods). `limit` limits total inflight request `tokens` from certain user or group and `request` reserves some amount of `tokens` for critical users or groups.
+
+
+## Motivation <a name="motivation"></a>
+
+Buggy service or malicious user may easily run out of '--max-mutating-requests-inflight' or '--max-requests-inflight' request limit by DOSing the apiserver. It will block actually important users (kubelets, controllers/schedulers), that, potentially, leads to impossibility to stop buggy service or block attacker via k8s.
+
+Using only `limit` to prevent blocking important users isn't enough flexible: to guarantee service availability for this users we should set rigid limits, when, sum of all limits not exceed server capacity (e.g. --max-requests-inflight). It will lead to throttling particular user or group even if the server not loaded by other users.
+
+Combination of `limit` and `request` allows to prevent overreservation of resources for critical users on the one hand, and flexible restrict different parts of the system from overspending resources on the other hand.
+
+Using limits in `tokens` instead of limiting number of requests allows to
+account resource consumption more realistic.
+
+
+### Goals <a name="goals"></a>
+
+* Implement per-user and per-group `limit` and `request`
+* Implement request cost estimation
+
+## Proposal <a name="proposal"></a>
+
+Implement weight-based per-user and per-group limits instead of current count-based limits.
+Every request has its own `weight` measured in `tokens`, according to its complexity. Exact algorithm of request cost calculation described below.
+
+Implement options `readonly_capacity`, `mutating_capacity`: Total `weight` of inflight requests that apiserver could carry.
+
+Implement new API resource `InflightLimit` to store group limits and requests:
+```
+apiVersion: extensions/v1beta1
+kind: InflightLimit
+metadata:
+  name: %NAME%
+spec:
+  subjects:
+    kind: %SUBJECT_KIND%
+    name: %SUBJECT_NAME%
+  readonly:
+    limit: %READONLY_TOKENS_LIMIT%
+    request: %READONLY_TOKENS_REQUEST%
+  mutating:
+    limit: %MUTATING_TOKENS_LIMIT%
+    request: %MUTATING_TOKENS_REQUEST%
+status:
+  readonlyInflightTokens: %CURRENT_READONLY_TOKENS_INFLIGHT%
+  mutatingInflightTokens: %CURRENT_MUTATING_TOKENS_INFLIGHT%
+```
+
+Where:
+
+`%NAME%` - The name of `InflightLimit` resource
+
+`%SUBJECT_KIND%` - `User` or `Group`, like in `RoleBinding`
+
+`%SUBJECT_NAME%` - User or group name
+
+`%READONLY_TOKENS_LIMIT%` - Limits total readonly requests `weight` across  particular user or group
+
+`%READONLY_TOKENS_REQUEST%` - Reserved (always available) `weight` of readonly requests  for particular user or group
```

Comment:
I think the `weight` is what I was trying to impose in a comment on the PR. Which is we need to be able to define a weight of a single request - we can treat "get single pod" the same as "get all pods from all namespaces".
Weight is supposed to somehow reflect that (though we need way more details about this in the KEP).

Reply:
> Reserved (always available ...

I think the bucket -> quota assignment should have multiple tiers:
* reserved (request)
* optimistic (limit) usage

We should have a good reason if we want to do something different than what we do for pod resource usage.

Reply:
@liggitt 
> what does "always available" mean? 

"Always available" means that given user/group always able to have at least `%READONLY_TOKENS_REQUEST%` inflight requests. E.g. if group `AdminGroup` has `%READONLY_TOKENS_REQUEST% = 10`, and there are 7 total inflight requests from users of this group, then given user, belonging to `AdminGroup`, should be able to make at least 3 more inflight requests, regardless of the current server load.

> what happens when the sum of the requests of all of the InflightLimit objects exceeds server capacity?

This is misconfiguration, we should avoid this cases.

> what does "weight" mean? does it mean a portion of the server capacity computed as server-capacity * (request / sum(all inflightlimit requests))?

As @wojtek-t  said, `weight` is a measure of a complexity of request for the server (see [#63719 comment](https://github.com/kubernetes/kubernetes/pull/63719#discussion_r195491243)). It's an abstract value, measured in the same units as `readonly_capacity` and `mutating_capacity`. 

> * how does the party defining the inflightlimit objects know what the server capacity is?

At current stage i have no answer, how to convert real CPU & RAM into exact value of `readonly_capacity` or `mutating_capacity`. I think, it should be clarified during discussion.

> the example below makes it sound as though the server capacity is expressed in the same units as the limits and requests, which makes it sound like overcommit is possible. if so:
> * what happens in cases of overcommit?

According to [Algorithm to accept/reject readonly request](https://github.com/kubernetes/community/blob/fa7602b2f9fb44b1f05adcd7dd9f2844e817ade5/keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md#algorithm-to-acceptreject-readonly-request-), in case of overcommit, we can guarantee that reserved requests (`%MUTATING_TOKENS_REQUEST%`, `%READONLY_TOKENS_REQUEST%`) will be processed. All requests above the reserved values will be processed as "first come first served": if server has available resources (i.e. total inflight `weight` less than `capacity`), request will be accepted, otherwise it will b rejected.